### DISCUSSION TITLE
Embedding smoldot into a mobile application

### DISCUSSION THREAD

**Discussion by finsig** [smol-dot/smoldot#600] [All]
I read the “Can I embed smoldot into a mobile application…” section of the readme and would like to embed smoldot-light into an iPhone app. Before I do that I would like to assess if it is possible. I do not (yet) have a good understanding of Wasm. My oversimplified belief is that smoldot compiles Rust to Wasm byte code, and then the web browser engine runs that bytecode. My worry is that the WebKit browser engine that is used for Safari, and by all iPhone apps, may have some technical limitation that makes this not possible. I notice, for example, that the Polkadot web extensions only support Chrome and Firefox. I realize that Safari is probably out of scope for your project, but do you know offhand of any potential issues with WebKit or have any online resources you would recommend I investigate?

**Comment by tomaka** [comment]
> My oversimplified belief is that smoldot compiles Rust to Wasm byte code, and then the web browser engine runs that bytecode.

That's correct!

> My worry is that the WebKit browser engine that is used for Safari, and by all iPhone apps, may have some technical limitation that makes this not possible.

There shouldn't be any problem with Safari: https://caniuse.com/wasm
WebAssembly is a pretty established technology by now.

The smoldot Wasm binary also uses some Wasm extensions such as SIMD, but I'm always careful to enable extensions only if they're widely supported according to the compatibility table here: https://webassembly.org/roadmap/
These extensions are enabled purely for optimizations, so if there's a popular browser that doesn't support a Wasm extension used by smoldot, I could easily disable it.

> I notice, for example, that the Polkadot web extensions only support Chrome and Firefox.

It supports only Chrome and Firefox because it's a browser extension. Writing a browser extension is different than writing just a web page.
Furthermore, I believe that it supports only Chrome and Firefox purely because nobody has bothered porting it to Safari, but technically I see no reason why this couldn't be done.
